plutus-tutorial/doc/plutus-tx.adoc

```diff
@@ -0,0 +1,383 @@
+integers.
+
+[source,haskell]
+----
+integerIdentity :: CompiledCode (Integer -> Integer)
+integerIdentity = $$(compile [|| \(x:: Integer) -> x ||])
+
+{- |
+>>> pretty $ getPlc integerIdentity
+(program 1.0.0
+  (lam ds (con integer) ds)
+)
+-}
+----
+
+== Functions and datatypes
+
+You can use functions inside your expression. In practice, you
+will usually want to define the entirety of your Plutus Tx program as a
+definition outside the quote, and then simply call it inside the quote.
+
+[source,haskell]
+----
+{-# INLINABLE plusOne #-} -- <.>
+plusOne :: Integer -> Integer
+plusOne x = x `addInteger` 1 -- <.>
+
+{-# INLINABLE myProgram #-}
+myProgram :: Integer
+myProgram =
+    let
+        plusOneLocal :: Integer -> Integer -- <.>
+        plusOneLocal x = x `addInteger` 1
+
+        localTwo = plusOneLocal 1
+        externalTwo = plusOne 1
+    in localTwo `addInteger` externalTwo
+
+functions :: CompiledCode Integer
+functions = $$(compile [|| myProgram ||])
+
+{- |
+>>> pretty $ runCk $ getPlc functions -- <.>
+(con 4)
+-}
+----
+<.> Functions which will be used in Plutus Tx programs should be
+marked with GHC’s `INLINABLE` pragma.
+This is usually necessary for non-local functions to
+be usable in Plutus Tx blocks, as it instructs GHC to keep the
+information that the Plutus Tx compiler needs. While you may be able
+to get away with omitting it, it is good practice to always include it.
+<.> Local functions do not need to be marked as `INLINABLE`.
+<.> `addInteger` comes from `Language.PlutusTx.Builtins`, and is
+mapped to the builtin integer addition function in Plutus Core.
+<.> We’ve used the CK evaluator for Plutus Core to evaluate the program
+and check that the result was what we expected
+
+We can use normal Haskell datatypes and pattern matching freely:
+
+[source,haskell]
+----
+matchMaybe :: CompiledCode (Maybe Integer -> Integer)
+matchMaybe = $$(compile [|| \(x:: Maybe Integer) -> case x of
+    Just n -> n
+    Nothing -> 0
+   ||])
+----
+
+Unlike functions, datatypes do not need any kind of special annotation to be
+used inside a
+quote, hence we can use types like `Maybe` from the Haskell `Prelude`.
+This works for your own datatypes too!
+
+Here’s a small example with a datatype of our own representing a
+potentially open-ended end date.
+
+[source,haskell]
+----
+-- | Either a specific end date, or "never".
+data EndDate = Fixed Integer | Never
+
+-- | Check whether a given time is past the end date.
+pastEnd :: CompiledCode (EndDate -> Integer -> Bool)
+pastEnd = $$(compile [|| \(end::EndDate) (current::Integer) -> case end of
+    Fixed n -> n `lessThanEqInteger` current
+    Never -> False
+   ||])
+----
+
+We could also have defined the `pastEnd` function as a separate `INLINABLE`
+binding and just referred to it in the quote, but in this case it's small enough
+to just write in place.
+
+== Typeclasses
+
+So far we have used functions like `lessThanEqInteger` for comparing `Integer` s,
```

Comment:
I think it might be better to use `+` instead of `addInteger` (etc.) from the beginning. The functions in the previous section are already different from regular simple Haskell functions (INLINE pragma, Template Haskell), and with `+` there would be one less source of confusion

Reply:
Well, we don't have `Num`, so we can't use `+`, but we could just use `<` off the bat, I guess. But given that I need to explain the builtins that appear in the generated Plutus Core I'm not sure we save much.

Reply:
true...